### PR TITLE
input/sqlserver: Add service and save connection pools

### LINKED ISSUE #5898: Reuse connections from connection pool in sqlserver input
### Proposal:

The sqlserver input should reuse connections from the connection pool.

### Current behavior:

New connection pool is created every interval.

### Desired behavior:

Hang on to the connection pool and reuse.

### Use case: 

Should be a lighter resource load on Telegraf and the server.

See also https://github.com/influxdata/telegraf/issues/3485#issuecomment-494887920

### PR DESCRIPTION
**Pull request by mjiderhamn**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [ ] Associated README.md updated.
- [x] Has appropriate unit tests.

Since the issue persists while #6713 was closed by author @nwneisen , I'm opening up a new PR to fix #5898.

### CODE REVIEW

**Comment by helenosheaa** [comment]
Thanks for reopening @mjiderhamn is this something you or @nwneisen are currently working on?

**Comment by mjiderhamn** [comment]
I'd be willing to try to take this to a mergeable state.

I have rebased onto `master` and made CircleCI happy. Anything else you're thinking about @helenosheaa ?

**Comment by helenosheaa** [comment]
@mjiderhamn sounds good, that's great! Nothing currently on my side.

**Comment by mjiderhamn** [comment]
Any news on getting this merged, @helenosheaa

**Comment by helenosheaa** [comment]
@mjiderhamn it looks like it has conflicts with master, if you can address those.

After that it needs to be reviewed, I'll take a look and request another reviewer :)

**Comment by mjiderhamn** [comment]
@helenosheaa `master` conflict resolved

**Comment by helenosheaa** [comment]
@denzilribeiro or @Trovalo would either of you mind taking a look at this?

**Comment by sjwang90** [comment]
@avinash-nigam would you or anyone else on your team be able to take a review on this PR?

File: plugins/inputs/sqlserver/sqlserver.go [hunk 1]
```diff
@@ -214,43 +212,56 @@ func initQueries(s *SQLServer) error {
...
+			}(pool, query)
 		}
 	}
 
+	wg.Wait()
+	return nil
+}
+
+// Start initialize a list of connection pools
+func (s *SQLServer) Start(acc telegraf.Accumulator) error {
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:232]: I see this func only used by unit test and not called from plugin. Is that intenional?

File: plugins/inputs/sqlserver/sqlserver.go [hunk 2]
```diff
@@ -15,14 +15,14 @@ import (
...
-	ExcludeQuery  []string `toml:"exclude_query"`
-	queries       MapQuery
-	isInitialized bool
+	Servers      []string `toml:"servers"`
+	QueryVersion int      `toml:"query_version"`
+	AzureDB      bool     `toml:"azuredb"`
+	DatabaseType string   `toml:"database_type"`
+	IncludeQuery []string `toml:"include_query"`
+	ExcludeQuery []string `toml:"exclude_query"`
+	pools        []*sql.DB
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserver.go:24]: AFAIK, for connection pooling, we usually have a default pool size and add connection to pool if the number of connections is less than pool size. was it there a specific reason of not having that mechanism? Is it safe to have any number?

**Comment by masree** [comment]
Looks good to me!

**Review by Jagularr — COMMENTED** [review]
Looks good! Would be curious to better understand how masree's comment might effect this but other than that this seems to make sense.

**Comment by mjiderhamn** [comment]
I did leave a reply to @masree , which is marked as "Pending" :-/ 

I wrote
> From [the docs](https://golang.org/pkg/database/sql/#DB) it seems to me you can only set max open vs idle, so I'm not sure what you mean with default pool size?
> 
> The idea is that each pool will only ever have 1 connection in them, since there is no concurrency involved. Yes, there should be 1 pool per connection string/server. The goal is to avoid opening and closing connections over and over for each interval, the way it currently does.

**Comment by helenosheaa** [comment]
@mjiderhamn could you address the merge conflicts again, then we can merge. Thanks for your work on this.

**Comment by masree** [comment]
> I did leave a reply to @masree , which is marked as "Pending" :-/
> 
> I wrote
> 
> > From [the docs](https://golang.org/pkg/database/sql/#DB) it seems to me you can only set max open vs idle, so I'm not sure what you mean with default pool size?
> > The idea is that each pool will only ever have 1 connection in them, since there is no concurrency involved. Yes, there should be 1 pool per connection string/server. The goal is to avoid opening and closing connections over and over for each interval, the way it currently does.

Sorry, I meant to say if we need to set max open and idle connections. Thanks for clarifying.

**Comment by mjiderhamn** [comment]
@helenosheaa I fixed the merge issues, but I don't understand what's up with CircleCI. If I run `make tidy` in an `quay.io/influxdb/telegraf-ci:1.15.5` container it is successful.

**Review by ssoroka — APPROVED** [review]

**Review by helenosheaa — APPROVED** [review]

**Pull request merged**